### DISCUSSION TITLE
Error Handling with ErrorPlugin

### DISCUSSION THREAD

**Discussion by furkandmrblk** [hayes/pothos#172] [Q&A] [answered]
Hey,
I am having trouble with implementing the error plugin into my project and wanted to ask what the error could be. I unfortunately couldn't solve it by myself. I am getting a response from my mutation that says:
```
message: "Expected returnType to be an object type"
__typename: "BaseError"
```

My builder looks like this:
```
export const builder = new SchemaBuilder<{
	... (defining some stuff here, has nothing to do with the error plugin tho)
}>({
	defaultInputFieldRequiredness: true,
	plugins: [
		SimpleObjectsPlugin,
		ScopeAuthPlugin,
		ValidationPlugin,
		SmartSubscriptionsPlugin,
		PrismaPlugin,
		ErrorsPlugin,
	],
	authScopes: async ({ user }) => ({
		public: true,
		user: !!user,
		unauthenticated: !user,
	}),
	smartSubscriptions: {
		...subscribeOptionsFromIterator((name, { pubsub }) => {
			return pubsub.asyncIterator(name);
		}),
	},
	prisma: { client: db },
	errorOptions: {
		defaultTypes: [Error],
	},
});
```
My error object & interface is the same as the one in the docs
```
const ErrorInterface = builder.interfaceRef<Error>('Error').implement({
	fields: (t) => ({
		message: t.exposeString('message'),
	}),
});

builder.objectType(Error, {
	name: 'BaseError',
	isTypeOf: (obj) => obj instanceof Error,
	interfaces: [ErrorInterface],
});
```
My problem is either in my resolver here or in my mutation on the frontend I suppose.
So here is my resolver, it's just for sending friend requests:

```
builder.mutationField('sendFriendRequest', (t) =>
	t.prismaField({
		type: db.friendRequest,
		args: {
			input: t.arg({ type: friendRequestInput }),
		},
		errors: {
			types: [Error],
		},
		resolve: async (_query, _root, { input }, { user, pubsub }) => {
			const friend = await friendRequestErrorCatcher(input.username, user!);

			const friendRequest = await db.friendRequest.findUnique({
				where: { userId: friend!.id },
			});

			if (friendRequest!.incomingUser.includes(user!.username))
				throw new Error('You already sent a friend request.');

			const friends: string[] = friendRequest!.incomingUser;

			friends.push(user!.username);

			const updatedFriendRequest = await db.friendRequest.update({
				where: { userId: friend!.id },
				data: {
					incomingUser: friends,
				},
			});

			await pubsub.publish('sent-friendrequest', updatedFriendRequest);
			return updatedFriendRequest;
		},
	}),
);
```
and the frontend mutation looks like this (I am using relay for the frontend) :
```
const [addFriend] = useMutation<FriendlistBlockAddFriendMutation>(graphql`
		mutation FriendlistBlockAddFriendMutation($input: friendRequestInput!) {
			sendFriendRequest(input: $input) {
				__typename
				... on Error {
					message
				}
				... on MutationSendFriendRequestSuccess {
					data {
						id
						incomingUser
					}
				}
			}
		}
	`);
```
the onSubmit for the form:

```
const onAdd = (e: any) => {
		e.preventDefault();
		addFriend({
			variables: {
				input: {
					username: username.username,
				},
			},
			onCompleted(response, err) {
				setSuccess(`Successfully sent request to ${username.username}`);
				setTimeout(() => {
					setSuccess(null);
				}, 2000);
				console.log('res', response); <---- Response output is the error described at the top.
				console.log('err', err);
			},
			onError(err) {
				console.log('onErr - err', err);
				setError(err.message);
				setTimeout(() => {
					setError(null);
				}, 2000);
			},
		});
	};
```

Thank you in advance and sorry to bother you guys with such a silly error 🙏

**Comment by furkandmrblk** [comment]
I just saw that I need to put the error plugin before the prisma plugin, so I did that. But it's still giving me the same error.

Updated builder:
```
export const builder = new SchemaBuilder<{
	...
}>({
	defaultInputFieldRequiredness: true,
	plugins: [
		ErrorsPlugin,
		SimpleObjectsPlugin,
		ScopeAuthPlugin,
		ValidationPlugin,
		SmartSubscriptionsPlugin,
		PrismaPlugin,
	],
	errorOptions: {
		defaultTypes: [Error],
	},
	authScopes: async ({ user }) => ({
		public: true,
		user: !!user,
		unauthenticated: !user,
	}),
	smartSubscriptions: {
		...subscribeOptionsFromIterator((name, { pubsub }) => {
			return pubsub.asyncIterator(name);
		}),
	},
	prisma: { client: db },
});
```

**Comment by hayes** [comment, marked as answer]
Can you update your plugins to the latest version? Looks like Prisma plugin
is using the old API.  The latest version of the Prisma plugin was a major
rewrite, and the way the integration with other plugins worked was
updated.  It's possible you have a new version of the errors plugin that
the old Prisma plugin doesn't know how to integrate with.

Not 100% sure this is the issue, but it's the most obvious thing I see.

**Reply by furkandmrblk** [reply]
Thanks, I updated everything. 
All my resolvers have lots of errors now, so I will need to fix those. I will come back when I fixed everything & look if the errors plugin is working now. 

Have a good day!

**Reply by hayes** [reply]
Yeah, breaking changes in the Prisma plugin are a bit annoying, but make it a lot better.  The main changes are the setup (follow the setup in the docs, since its very different) and models are now referenced by string now ('User' instead of db.user)

**Reply by hayes** [reply]
Yeah, breaking changes in the Prisma plugin are a bit annoying, but make it a lot better.  The main changes are the setup (follow the setup in the docs, since its very different) and models are now referenced by string now ('User' instead of db.user)

**Reply by furkandmrblk** [reply]
Yup, thank you. I will look into it now.

**Comment by furkandmrblk** [comment]
Sorry, I had to do some things and fixed everything in my backend now. It was indeed because of the old prisma plugin version!
The only thing that's kind of weird now is, that the error gets categorised as a response and not an error. But that's fine for now.

The solution was to keep your packages up to date!

Thank you again for your help.

**Comment by hayes** [comment]
The whole point of the plugin is to make errors part of the normal
response, so this is expected behavior.  GraphQL in general will always
return a 200 even if fields in the response throw errors.  If you wanted to
change this it could probably be done with graphql helix and a custom http
handler that checked a response for errors before setting the status code.

**Reply by furkandmrblk** [reply]
Thank you, learned something new! I don't know how to close this discussion, because I rarely ask questions, sorry.
